File: src/main/java/com/baulsupp/oksocial/services/spotify/SpotifyAuthInterceptor.java
package com.baulsupp.oksocial.services.spotify;

import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.authenticator.AuthUtil;
import com.baulsupp.oksocial.authenticator.BasicCredentials;
import com.baulsupp.oksocial.authenticator.JsonCredentialsValidator;
import com.baulsupp.oksocial.authenticator.ValidatedCredentials;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2ServiceDefinition;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token;
import com.baulsupp.oksocial.completion.ApiCompleter;
import com.baulsupp.oksocial.completion.BaseUrlCompleter;
import com.baulsupp.oksocial.completion.CompletionVariableCache;
import com.baulsupp.oksocial.completion.UrlList;
import com.baulsupp.oksocial.credentials.CredentialsStore;
import com.baulsupp.oksocial.secrets.Secrets;
import com.baulsupp.oksocial.services.lyft.LyftUtil;
import com.google.common.collect.Sets;
import com.baulsupp.oksocial.output.OutputHandler;
import java.io.IOException;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.Future;
import okhttp3.Credentials;
import okhttp3.FormBody;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import static java.util.Optional.of;

public class SpotifyAuthInterceptor implements AuthInterceptor<Oauth2Token> {
  @Override public Oauth2ServiceDefinition serviceDefinition() {
    return new Oauth2ServiceDefinition(host(), "Spotify API", "spotify",
        "https://developer.spotify.com/web-api/endpoint-reference/",
        "https://developer.spotify.com/my-applications/");
  }

  protected String host() {
    return "api.spotify.com";
  }

  @Override public Response intercept(Interceptor.Chain chain, Oauth2Token credentials)
      throws IOException {
    Request request = chain.request();

    String token = credentials.accessToken;

    request =
        request.newBuilder().addHeader("Authorization", "Bearer " + token).build();

    return chain.proceed(request);
  }

  @Override public Oauth2Token authorize(OkHttpClient client, OutputHandler outputHandler,
      List<String> authArguments) throws IOException {
    System.err.println("Authorising Spotify API");

    String clientId =
        Secrets.prompt("Spotify Client Id", "spotify.clientId", "", false);
    String clientSecret =
        Secrets.prompt("Spotify Client Secret", "spotify.clientSecret", "", true);

    Set<String> scopes =
        Secrets.promptArray("Scopes", "spotify.scopes", SpotifyUtil.SCOPES);

    return SpotifyAuthFlow.login(client, outputHandler, clientId, clientSecret, scopes);
  }

  @Override public ApiCompleter apiCompleter(String prefix, OkHttpClient client,
      CredentialsStore credentialsStore, CompletionVariableCache completionVariableCache)
      throws IOException {
    return new BaseUrlCompleter(UrlList.fromResource(name()).get(), hosts());
  }

  @Override public Future<Optional<ValidatedCredentials>> validate(OkHttpClient client,
      Request.Builder requestBuilder, Oauth2Token credentials) throws IOException {
    return new JsonCredentialsValidator(
        SpotifyUtil.apiRequest("/v1/me", requestBuilder),
        map -> "" + map.get("display_name")).validate(client);
  }

  @Override public Collection<String> hosts() {
    return Collections.unmodifiableSet(Sets.newHashSet(
        "api.spotify.com")
    );
  }

  @Override public boolean canRenew(Oauth2Token credentials) {
    return credentials.refreshToken.isPresent()
        && credentials.clientId.isPresent()
        && credentials.clientSecret.isPresent();
  }

  @Override public Optional<Oauth2Token> renew(OkHttpClient client, Oauth2Token credentials)
      throws IOException {
    String tokenUrl = "https://accounts.spotify.com/api/token";

    RequestBody body =
        new FormBody.Builder()
            .add("refresh_token", credentials.refreshToken.get())
            .add("grant_type", "refresh_token")
            .build();

    Request request = new Request.Builder().header("Authorization",
        Credentials.basic(credentials.clientId.get(), credentials.clientSecret.get()))
        .url(tokenUrl)
        .method("POST", body)
        .build();

    Map<String, Object> responseMap = AuthUtil.makeJsonMapRequest(client, request);

    return of(new Oauth2Token((String) responseMap.get("access_token"),
        (String) responseMap.get("refresh_token"), credentials.clientId.get(),
        credentials.clientSecret.get()));
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/services/spotify/SpotifyAuthInterceptor.kt
package com.baulsupp.oksocial.services.spotify

import com.baulsupp.oksocial.Token
import com.baulsupp.oksocial.TokenValue
import com.baulsupp.oksocial.authenticator.AuthInterceptor
import com.baulsupp.oksocial.authenticator.AuthUtil
import com.baulsupp.oksocial.authenticator.ValidatedCredentials
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2ServiceDefinition
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token
import com.baulsupp.oksocial.completion.ApiCompleter
import com.baulsupp.oksocial.completion.BaseUrlCompleter
import com.baulsupp.oksocial.completion.CompletionVariableCache
import com.baulsupp.oksocial.completion.UrlList
import com.baulsupp.oksocial.credentials.CredentialsStore
import com.baulsupp.oksocial.kotlin.moshi
import com.baulsupp.oksocial.kotlin.queryMapValue
import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.oksocial.secrets.Secrets
import com.baulsupp.oksocial.services.spotify.model.ErrorResponse
import com.baulsupp.oksocial.util.ClientException
import okhttp3.Credentials
import okhttp3.FormBody
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.util.Arrays

class SpotifyAuthInterceptor : AuthInterceptor<Oauth2Token>() {
  override fun serviceDefinition(): Oauth2ServiceDefinition {
    return Oauth2ServiceDefinition("api.spotify.com", "Spotify API", "spotify",
      "https://developer.spotify.com/web-api/endpoint-reference/",
      "https://developer.spotify.com/my-applications/")
  }

  override fun intercept(chain: Interceptor.Chain, credentials: Oauth2Token): Response {
    var request = chain.request()

    val token = credentials.accessToken

    request = request.newBuilder().addHeader("Authorization", "Bearer " + token).build()

    return chain.proceed(request)
  }

  override suspend fun authorize(client: OkHttpClient, outputHandler: OutputHandler<Response>,
                                 authArguments: List<String>): Oauth2Token {

    val clientId = Secrets.prompt("Spotify Client Id", "spotify.clientId", "", false)
    val clientSecret = Secrets.prompt("Spotify Client Secret", "spotify.clientSecret", "", true)

    val scopes = Secrets.promptArray("Scopes", "spotify.scopes",
      Arrays.asList("playlist-read-private",
        "playlist-read-collaborative",
        "playlist-modify-public",
        "playlist-modify-private",
        "streaming",
        "ugc-image-upload",
        "user-follow-modify",
        "user-follow-read",
        "user-library-read",
        "user-library-modify",
        "user-read-private",
        "user-read-birthdate",
        "user-read-email",
        "user-top-read"))

    return SpotifyAuthFlow.login(client, outputHandler, clientId, clientSecret, scopes)
  }

  override fun apiCompleter(prefix: String, client: OkHttpClient,
                            credentialsStore: CredentialsStore,
                            completionVariableCache: CompletionVariableCache,
                            tokenSet: Token): ApiCompleter {
    return BaseUrlCompleter(UrlList.fromResource(name())!!, hosts(), completionVariableCache)
  }

  override suspend fun validate(client: OkHttpClient,
                                credentials: Oauth2Token): ValidatedCredentials {
    return ValidatedCredentials(client.queryMapValue<String>("https://api.spotify.com/v1/me", TokenValue(credentials), "display_name"))
  }

  override fun hosts(): Set<String> = setOf("api.spotify.com")

  override fun canRenew(credentials: Oauth2Token): Boolean = credentials.isRenewable()

  override fun errorMessage(ce: ClientException): String {
    if (ce.code == 401) {
      try {
        val message = ce.responseMessage

        return moshi.adapter(ErrorResponse::class.java).fromJson(message)!!.error.message
      } catch (e: Exception) {
        // ignore
      }
    }

    return super.errorMessage(ce)
  }

  override suspend fun renew(client: OkHttpClient, credentials: Oauth2Token): Oauth2Token? {
    val tokenUrl = "https://accounts.spotify.com/api/token"

    val body = FormBody.Builder()
      .add("refresh_token", credentials.refreshToken!!)
      .add("grant_type", "refresh_token")
      .build()

    val request = Request.Builder().header("Authorization",
      Credentials.basic(credentials.clientId!!, credentials.clientSecret!!))
      .url(tokenUrl)
      .method("POST", body)
      .build()

    val responseMap = AuthUtil.makeJsonMapRequest(client, request)

    return Oauth2Token(responseMap["access_token"] as String,
      responseMap["refresh_token"] as String?, credentials.clientId,
      credentials.clientSecret)
  }
}
